homeassistant/components/splunk/__init__.py

```diff
@@ -95,19 +120,52 @@ def splunk_event_listener(event):
         except ValueError:
             _state = state.state
 
-        json_body = [
+        queue(
             {
-                "domain": state.domain,
-                "entity_id": state.object_id,
-                "attributes": dict(state.attributes),
-                "time": str(event.time_fired),
-                "value": _state,
+                "time": event.time_fired.timestamp(),
                 "host": name,
+                "event": {
+                    "domain": state.domain,
+                    "entity_id": state.object_id,
+                    "attributes": dict(state.attributes),
+                    "value": _state,
+                },
             }
-        ]
-
-        post_request(event_collector, json_body, headers, verify_ssl)
+        )
+        return True
+
+    def splunk_send(event=None):
+        """Send batched messages to Splunk every second."""
+        nonlocal batch
+        if len(batch) > 0:
+            events = batch
+            batch = []
+            try:
+                event_collector._send_to_splunk("send-event", "".join(events))
+            except (
+                request_exceptions.HTTPError,
+                request_exceptions.Timeout,
+                request_exceptions.ConnectionError,
+                request_exceptions.TooManyRedirects,
+            ) as err:
+                _LOGGER.warning(err)
+
+    def splunk_stop(event):
+        queue(
+            {
+                "time": event.time_fired.timestamp(),
+                "host": name,
+                "event": {
+                    "domain": DOMAIN,
+                    "meta": "Home Assistant is stopping",
+                    "value": 0,
+                },
+            }
+        )
+        splunk_send()
 
     hass.bus.listen(EVENT_STATE_CHANGED, splunk_event_listener)
+    hass.bus.listen(EVENT_TIME_CHANGED, splunk_send)
```

Comment:
Do we want to potentially send every second? That's quite frequent.

Reply:
So without this, Home Assistant was making a large number of simultaneous connections, which can cause issues for Splunk, and exceeded the requests connection pool.

With this, it sends at most 1 payload per second, and when there is no data to send, the if statement fails and nothing happens.

Reply:
The batch feature is good. I'm just thinking if we want to send the batch at a lower frequency than 1 Hz?

We have helpers to schedule functions at time intervals in the event helper module.

Reply:
Eg every 30 seconds. Is there a limit to how large the payload can be?

Reply:
512KB is the payload size limit, so that should be high enough to not cause issues. Personally I think 30s is too long, it would completely break any real-time graphing in Splunk, but something in the range of 1-10 seconds could be acceptable.

To apply some statistical analysis to this, I checked my past Splunk events from Home Assistant, and I have been recording an average of 1.3 events per second, and if you remove seconds where no state changes occurred (so the if statement would cause nothing to happen) that increases to an average of 4.9 events per second. So even at 1Hz it should be a reduction in transactions.

My understanding is the way its implemented should only use a few cpu cycles to fail the if statement, but if its a concern I'll happily look into the scheduling helpers.

Reply:
A 10 second interval sounds good.

Reply:
I've come up with a better solution that doesn't depend on timers. It batches events only while a send is currently in progress. So basically its enforcing a single threat of HTTP POST to Splunk, which wont exceed a connection pool, wont overload the server, and will do absolutely nothing in Home Assistant when there are no events to send.